x-pack/plugin/write-load-forecaster/src/main/java/org/elasticsearch/xpack/writeloadforecaster/LicensedWriteLoadForecaster.java

```diff
@@ -0,0 +1,145 @@
+        TimeValue.timeValueMinutes(1),
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    private final BooleanSupplier hasValidLicense;
+    private volatile TimeValue maxIndexAge;
+    private volatile TimeValue minShardUptime;
+
+    LicensedWriteLoadForecaster(BooleanSupplier hasValidLicense, Settings settings, ClusterSettings clusterSettings) {
+        this(hasValidLicense, MAX_INDEX_AGE_SETTING.get(settings), MIN_UPTIME_SETTING.get(settings));
+        clusterSettings.addSettingsUpdateConsumer(MAX_INDEX_AGE_SETTING, this::setMaxIndexAgeSetting);
+        clusterSettings.addSettingsUpdateConsumer(MIN_UPTIME_SETTING, this::setMinUptimeSetting);
+    }
+
+    LicensedWriteLoadForecaster(BooleanSupplier hasValidLicense, TimeValue maxIndexAge, TimeValue minShardUptime) {
+        this.hasValidLicense = hasValidLicense;
+        this.maxIndexAge = maxIndexAge;
+        this.minShardUptime = minShardUptime;
+    }
+
+    private void setMaxIndexAgeSetting(TimeValue updatedMaxIndexAge) {
+        this.maxIndexAge = updatedMaxIndexAge;
+    }
+
+    private void setMinUptimeSetting(TimeValue updatedMinShardUptime) {
+        this.minShardUptime = updatedMinShardUptime;
+    }
+
+    @Override
+    public ClusterState withWriteLoadForecastForWriteIndex(String dataStreamName, ClusterState clusterState) {
+        if (hasValidLicense.getAsBoolean() == false) {
+            return clusterState;
+        }
+
+        DataStream dataStream = clusterState.metadata().dataStreams().get(dataStreamName);
+
+        if (dataStream == null) {
+            return clusterState;
+        }
+
+        double maxWriteLoadAvg = Double.MIN_VALUE;
+        double totalWriteLoadAvg = 0;
+        double totalIndices = 0;
+
+        for (Index index : dataStream.getIndices()) {
+            final IndexMetadata indexMetadata = clusterState.metadata().getIndexSafe(index);
+            final long indexAge = System.currentTimeMillis() - indexMetadata.getCreationDate();
+            final IndexWriteLoad writeLoad = indexMetadata.getWriteLoad();
+
+            if (index.equals(dataStream.getWriteIndex()) || indexAge > maxIndexAge.millis() || writeLoad == null) {
+                continue;
+            }
+
+            double totalIndexWriteLoad = 0;
+            int totalShards = 0;
+            for (int shardId = 0; shardId < indexMetadata.getNumberOfShards(); shardId++) {
+                final OptionalDouble writeLoadForShard = writeLoad.getWriteLoadForShard(shardId);
+                final OptionalLong uptimeInMillisForShard = writeLoad.getUptimeInMillisForShard(shardId);
+                if (writeLoadForShard.isPresent()) {
+                    assert uptimeInMillisForShard.isPresent();
+                    double shardWriteLoad = writeLoadForShard.getAsDouble();
+                    long shardUptimeInMillis = uptimeInMillisForShard.getAsLong();
+                    if (shardUptimeInMillis > minShardUptime.millis()) {
```

Comment:
I wonder if we can improve this, avoiding the min shard up time setting. My idea is to use the up-time to weigh the avg. i.e., we compute sum(shardWriteLoad*shardUptimeInMillis) and sum(shardUpTimeInMillis) (summing over all shards of all indices) and then calculate the final `indexWriteLoadAvg` (a few lines below) as

`sum(shardWriteLoad*shardUptimeInMillis) / sum(shardUpTimeInMillis)`

That way, if a shard had very short uptime, it's impact on the `indexWriteLoadAvg` will also be very small. And we avoid the minimum shard uptime setting, since the inaccuracy of short uptimes will not be important, but we'll still use whatever we have as the basis.

Otherwise I think we still risk not having any data if we rollover for instance every 30 mins on an index (unless if I missed where that is addressed).